### ISSUE TITLE
Combined model and jacobian function

### ISSUE THREAD

**Issue by asprionj** [JuliaNLSolvers/LsqFit.jl#196]
In some problems, it is efficient to calculate the model (i.e. trajectory) along with the jacobian (i.e. gradients of the trajectory w.r.t. the parameters), at once. If everything goes fine (full Gauss-Newton steps), each iteration requires exactly those two pieces of information, correct? It don't know the internals; would adding the possibility of specifying a function that evaluates both at the same time be possible?

**Comment by pkofod** [comment]
Yes, this is possible in Optim so should also be possible here. Sometimes it can depend a bit on the way the algorithm has been written, but in Optim for example I've allowed for the trust region code (similar to LM) to have a switch for whether to evaluate together or not (which is better depends on whether the objective is cheap to evaluate with the gradient or not)

**Comment by asprionj** [comment]
Is this feature documented somewhere? (Specifically for LsqFit) Otherwise I can come up with a simple example to document it and illustrate the possible benefits.
